src/utils.js

```diff
@@ -0,0 +1,174 @@
+import axios from 'axios/index';
+import crypto from 'crypto';
+import ReactHtmlParser, { convertNodeToElement } from 'react-html-parser';
+import { serialize } from 'react-serialize';
+
+// Minimal logging module mimicking Python logging module
+export const logging = {
+  CRITICAL: 50,
+  ERROR: 40,
+  WARNING: 30,
+  INFO: 20,
+  DEBUG: 10,
+  getLogger: (logLevel, out = console) => {
+    return {
+      critical: msg =>
+        logging.CRITICAL >= logLevel ? out.error(`Plone – ${msg}`) : null,
+      error: msg =>
+        logging.ERROR >= logLevel ? out.error(`Plone – ${msg}`) : null,
+      warn: msg =>
+        logging.WARNING >= logLevel ? out.warn(`Plone – ${msg}`) : null,
+      info: msg =>
+        logging.INFO >= logLevel ? out.info(`Plone – ${msg}`) : null,
+      debug: msg =>
+        logging.DEBUG >= logLevel ? out.debug(`Plone – ${msg}`) : null,
+    };
+  },
+};
+
+// Create content digest (hash)
+export const createContentDigest = data =>
+  crypto
+    .createHash('md5')
+    .update(JSON.stringify(data))
+    .digest('hex');
+
+// Get URL without query parameters
+export const urlWithoutParameters = url => {
+  return url.split('?')[0];
+};
+
+// Return parent id for Plone REST API content object id
+export const parentId = id =>
+  id.match(/\//g).length > 3
+    ? id
+        .split('/')
+        .slice(0, -1)
+        .join('/')
+    : id;
+
+// Normalize path
+export const normalizePath = path => {
+  return path ? path.replace(/^\/*/, '/').replace(/\/*$/, '/') : '/';
+};
+
+// Add token to header when given
+export const headersWithToken = (headers, token) =>
+  token ? { ...headers, Authorization: `Bearer ${token}` } : headers;
+
+// Fetch JSON data from an URL
+export const fetchUrl = async (url, token, params, http = axios) => {
+  const response = await http.get(url, {
+    headers: headersWithToken(
+      {
+        accept: 'application/json',
+      },
+      token
+    ),
+    params: params,
+  });
+  return response.data || {};
+};
+
+// Fetch Plone JSON REST API data with batching expanded
+export const fetchPlone = async (url, token, params, http = axios) => {
+  let batch, data;
+  batch = data = await fetchUrl(url, token, params, http);
+  while (1) {
+    if (batch['batching']) {
+      if (batch['batching'].next) {
+        batch = await fetchUrl(batch['batching'].next, token, {}, http);
+        data.items.push(...batch.items);
+      } else break;
+    } else {
+      break;
+    }
+  }
+  return data;
+};
+
+// Normalize Plone JSON to be usable as such in GatsbyJS
+export const normalizeData = function(data, baseUrl) {
```

Comment:
@ajayns I moved a lot of data processing logic from `gatsby-node.js` into this function so that it would be easier to write tests for it.

Reply:
Understood, I think we have a way more cleaner code base now!